File: expo/android/src/main/java/expo/modules/autograph/ExpoAutographModule.kt
package expo.modules.autograph

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import sh.autograph.authenticate
import sh.autograph.certify
import sh.autograph.createNonce
import sh.autograph.createSkippedIndexes
import sh.autograph.decrypt
import sh.autograph.encrypt
import sh.autograph.generateIdentityKeyPair
import sh.autograph.generateSecretKey
import sh.autograph.generateSessionKeyPair
import sh.autograph.getIdentityPublicKey
import sh.autograph.getSessionPublicKey
import sh.autograph.keyExchange
import sh.autograph.ready
import sh.autograph.verify
import sh.autograph.verifyKeyExchange

class ExpoAutographModule : Module() {
    override fun definition() =
        ModuleDefinition {
            Name("ExpoAutograph")

            Events("onReady")

            OnStartObserving {
                try {
                    ready()
                    sendEvent("onReady")
                } catch (e: Exception) {
                }
            }

            Function("authenticate") { ourIdentityKeyPair: ByteArray, theirIdentityKey: ByteArray ->
                try {
                    val safetyNumber =
                        authenticate(
                            ourIdentityKeyPair,
                            byteArrayOf(1, 2, 3),
                            theirIdentityKey,
                            byteArrayOf(4, 5, 6),
                        )
                    mapOf(
                        "success" to true,
                        "safetyNumber" to safetyNumber,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "safetyNumber" to ByteArray(0),
                    )
                }
            }

            Function("certify") { ourIdentityKeyPair: ByteArray, theirIdentityKey: ByteArray, data: ByteArray ->
                try {
                    val signature =
                        certify(
                            ourIdentityKeyPair,
                            theirIdentityKey,
                            data,
                        )
                    mapOf(
                        "success" to true,
                        "signature" to signature,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "signature" to ByteArray(0),
                    )
                }
            }

            Function("verify") { ownerIdentityKey: ByteArray, certifierIdentityKey: ByteArray, signature: ByteArray, data: ByteArray ->
                verify(
                    ownerIdentityKey,
                    certifierIdentityKey,
                    signature,
                    data,
                )
            }

            Function(
                "keyExchange",
            ) {
                    isInitiator: Boolean,
                    ourIdentityKeyPair: ByteArray,
                    ourSessionKeyPair: ByteArray,
                    theirIdentityKey: ByteArray,
                    theirSessionKey: ByteArray,
                ->
                try {
                    val (transcript, ourSignature, sendingKey, receivingKey) =
                        keyExchange(
                            isInitiator,
                            ourIdentityKeyPair,
                            ourSessionKeyPair,
                            theirIdentityKey,
                            theirSessionKey,
                        )
                    mapOf(
                        "success" to true,
                        "transcript" to transcript,
                        "ourSignature" to ourSignature,
                        "sendingKey" to sendingKey,
                        "receivingKey" to receivingKey,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "transcript" to ByteArray(0),
                        "ourSignature" to ByteArray(0),
                        "sendingKey" to ByteArray(0),
                        "receivingKey" to ByteArray(0),
                    )
                }
            }

            Function(
                "verifyKeyExchange",
            ) { transcript: ByteArray, ourIdentityKeyPair: ByteArray, theirIdentityKey: ByteArray, theirSignature: ByteArray ->
                try {
                    verifyKeyExchange(
                        transcript,
                        ourIdentityKeyPair,
                        theirIdentityKey,
                        theirSignature,
                    )
                    true
                } catch (e: Exception) {
                    false
                }
            }

            Function("generateIdentityKeyPair") {
                try {
                    val keyPair = generateIdentityKeyPair()
                    mapOf(
                        "success" to true,
                        "keyPair" to keyPair,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "keyPair" to ByteArray(0),
                    )
                }
            }

            Function("generateSessionKeyPair") {
                try {
                    val keyPair = generateSessionKeyPair()
                    mapOf(
                        "success" to true,
                        "keyPair" to keyPair,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "keyPair" to ByteArray(0),
                    )
                }
            }

            Function("getIdentityPublicKey") { keyPair: ByteArray ->
                getIdentityPublicKey(keyPair)
            }

            Function("getSessionPublicKey") { keyPair: ByteArray ->
                getSessionPublicKey(keyPair)
            }

            Function("getPublicKeys") { identityKeyPair: ByteArray, sessionKeyPair: ByteArray ->
                val identityKey = getIdentityPublicKey(identityKeyPair)
                val sessionKey = getSessionPublicKey(sessionKeyPair)
                mapOf(
                    "identityKey" to identityKey,
                    "sessionKey" to sessionKey,
                )
            }

            Function("createNonce") {
                createNonce()
            }

            Function("createSkippedIndexes") { count: Int ->
                createSkippedIndexes(count)
            }

            Function("generateSecretKey") {
                try {
                    val key = generateSecretKey()
                    mapOf(
                        "success" to true,
                        "key" to key,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "key" to ByteArray(0),
                    )
                }
            }

            Function("encrypt") { key: ByteArray, nonce: ByteArray, plaintext: ByteArray ->
                try {
                    val (index, ciphertext) =
                        encrypt(
                            key,
                            nonce,
                            plaintext,
                        )
                    mapOf(
                        "success" to true,
                        "nonce" to nonce.copyOf(),
                        "index" to index,
                        "ciphertext" to ciphertext,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "nonce" to ByteArray(0),
                        "index" to 0,
                        "ciphertext" to ByteArray(0),
                    )
                }
            }

            Function("decrypt") { key: ByteArray, nonce: ByteArray, skippedIndexes: ByteArray, ciphertext: ByteArray ->
                try {
                    val (index, plaintext) =
                        decrypt(
                            key,
                            nonce,
                            skippedIndexes,
                            ciphertext,
                        )
                    mapOf(
                        "success" to true,
                        "nonce" to nonce.copyOf(),
                        "skippedIndexes" to skippedIndexes.copyOf(),
                        "index" to index,
                        "plaintext" to plaintext,
                    )
                } catch (e: Exception) {
                    mapOf(
                        "success" to false,
                        "nonce" to ByteArray(0),
                        "skippedIndexes" to ByteArray(0),
                        "index" to 0,
                        "plaintext" to ByteArray(0),
                    )
                }
            }
        }
}

File: expo/ios/ExpoAutographModule.swift
import Autograph
import ExpoModulesCore
import Foundation

extension Data {
    func toBytes() -> [UInt8] {
        [UInt8](self)
    }
}

public class ExpoAutographModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoAutograph")

        Events("onReady")

        OnStartObserving {
            do {
                try Autograph.ready()
                sendEvent("onReady")
            } catch {}
        }

        Function("authenticate") {
            (ourIdentityKeyPair: Data, theirIdentityKey: Data) -> [String: Any?] in
            do {
                let safetyNumber = try Autograph.authenticate(
                    ourIdentityKeyPair.toBytes(),
                    [1, 2, 3],
                    theirIdentityKey.toBytes(),
                    [4, 5, 6]
                )
                return [
                    "success": true,
                    "safetyNumber": Data(safetyNumber),
                ]
            } catch {
                return [
                    "success": false,
                    "safetyNumber": Data(),
                ]
            }
        }

        Function("certify") {
            (ourIdentityKeyPair: Data, theirIdentityKey: Data, data: Data) -> [String: Any?] in
            do {
                let signature = try Autograph.certify(
                    ourIdentityKeyPair.toBytes(),
                    theirIdentityKey.toBytes(),
                    data.toBytes()
                )
                return ["success": true, "signature": Data(signature)]
            } catch {
                return [
                    "success": false,
                    "signature": Data(),
                ]
            }
        }

        Function("verify") {
            (ownerIdentityKey: Data, certifierIdentityKey: Data, signature: Data, data: Data) in
            Autograph.verify(
                ownerIdentityKey.toBytes(),
                certifierIdentityKey.toBytes(),
                signature.toBytes(),
                data.toBytes()
            )
        }

        Function("keyExchange") {
            (
                isInitiator: Bool,
                ourIdentityKeyPair: Data,
                ourSessionKeyPair: Data,
                theirIdentityKey: Data,
                theirSessionKey: Data
            ) -> [String: Any?] in
            do {
                let (
                    transcript,
                    ourSignature,
                    sendingKey,
                    receivingKey
                ) = try Autograph.keyExchange(
                    isInitiator,
                    ourIdentityKeyPair.toBytes(),
                    ourSessionKeyPair.toBytes(),
                    theirIdentityKey.toBytes(),
                    theirSessionKey.toBytes()
                )
                return [
                    "success": true,
                    "transcript": Data(transcript),
                    "ourSignature": Data(ourSignature),
                    "sendingKey": Data(sendingKey),
                    "receivingKey": Data(receivingKey),
                ]
            } catch {
                return [
                    "success": false,
                    "transcript": Data(),
                    "ourSignature": Data(),
                    "sendingKey": Data(),
                    "receivingKey": Data(),
                ]
            }
        }

        Function("verifyKeyExchange") {
            (
                transcript: Data,
                ourIdentityKeyPair: Data,
                theirIdentityKey: Data,
                theirSignature: Data
            ) -> Bool in
            do {
                try Autograph.verifyKeyExchange(
                    transcript.toBytes(),
                    ourIdentityKeyPair.toBytes(),
                    theirIdentityKey.toBytes(),
                    theirSignature.toBytes()
                )
                return true
            } catch {
                return false
            }
        }

        Function("generateIdentityKeyPair") { () -> [String: Any?] in
            do {
                let keyPair = try Autograph.generateIdentityKeyPair()
                return [
                    "success": true,
                    "keyPair": Data(keyPair),
                ]
            } catch {
                return [
                    "success": false,
                    "keyPair": Data(),
                ]
            }
        }

        Function("generateSessionKeyPair") { () -> [String: Any?] in
            do {
                let keyPair = try Autograph.generateSessionKeyPair()
                return [
                    "success": true,
                    "keyPair": Data(keyPair),
                ]
            } catch {
                return [
                    "success": false,
                    "keyPair": Data(),
                ]
            }
        }

        Function("getIdentityPublicKey") { (keyPair: Data) -> Data in
            let publicKey = Autograph.getIdentityPublicKey(keyPair.toBytes())
            return Data(publicKey)
        }

        Function("getSessionPublicKey") { (keyPair: Data) -> Data in
            let publicKey = Autograph.getSessionPublicKey(keyPair.toBytes())
            return Data(publicKey)
        }

        Function("getPublicKeys") {
            (identityKeyPair: Data, sessionKeyPair: Data) -> [String: Data] in
            let identityKey =
                Autograph
                .getIdentityPublicKey(identityKeyPair.toBytes())
            let sessionKey =
                Autograph
                .getSessionPublicKey(sessionKeyPair.toBytes())
            return [
                "identityKey": Data(identityKey),
                "sessionKey": Data(sessionKey),
            ]
        }

        Function("createNonce") { () -> Data in
            let nonce = Autograph.createNonce()
            return Data(nonce)
        }

        Function("createSkippedIndexes") { (count: UInt16) -> Data in
            let indexes = Autograph.createSkippedIndexes(count)
            return Data(indexes)
        }

        Function("generateSecretKey") { () -> [String: Any?] in
            do {
                let key = try Autograph.generateSecretKey()
                return [
                    "success": true,
                    "key": Data(key),
                ]
            } catch {
                return [
                    "success": false,
                    "key": Data(),
                ]
            }
        }

        Function("encrypt") { (key: Data, nonce: Data, plaintext: Data) -> [String: Any?] in
            do {
                var n = nonce.toBytes()
                let (index, ciphertext) = try Autograph.encrypt(
                    key.toBytes(),
                    &n,
                    plaintext.toBytes()
                )
                return [
                    "success": true,
                    "nonce": Data(n),
                    "index": index,
                    "ciphertext": Data(ciphertext),
                ]
            } catch {
                return [
                    "success": false,
                    "nonce": Data(),
                    "index": 0,
                    "ciphertext": Data(),
                ]
            }
        }

        Function("decrypt") {
            (key: Data, nonce: Data, skippedIndexes: Data, ciphertext: Data) -> [String: Any?] in
            do {
                var n = nonce.toBytes()
                var indexes = skippedIndexes.toBytes()
                let (index, plaintext) = try Autograph.decrypt(
                    key.toBytes(),
                    &n,
                    &indexes,
                    ciphertext.toBytes()
                )
                return [
                    "success": true,
                    "nonce": Data(n),
                    "skippedIndexes": Data(indexes),
                    "index": index,
                    "plaintext": Data(plaintext),
                ]
            } catch {
                return [
                    "success": false,
                    "nonce": Data(),
                    "skippedIndexes": Data(),
                    "index": 0,
                    "plaintext": Data(),
                ]
            }
        }
    }
}
